sql/2024/04/image-formats.sql

```diff
@@ -0,0 +1,143 @@
+# HTTP Archive query to get median image size per image type for LCP images.
+#
+# WPP Research, Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# See https://github.com/GoogleChromeLabs/wpp-research/pull/97
+
+DECLARE
+  DATE_TO_QUERY DATE DEFAULT '2024-03-01';
+
+CREATE TEMPORARY FUNCTION
+  IS_CMS(technologies ARRAY<STRUCT<technology STRING,
+                                   categories ARRAY<STRING>,
+                                   info ARRAY<STRING>>>,
+         cms STRING,
+         version STRING)
+  RETURNS BOOL AS ( EXISTS(
+  SELECT
+    *
+  FROM
+    UNNEST(technologies) AS technology,
+    UNNEST(technology.info) AS info
+  WHERE
+      technology.technology = cms
+    AND ( version = ""
+    OR ENDS_WITH(version, ".x")
+            AND (STARTS_WITH(info, RTRIM(version, "x"))
+        OR info = RTRIM(version, ".x"))
+    OR info = version ) ) );
+
+CREATE TEMPORARY FUNCTION
+  IS_IMAGE (summary STRING)
+  RETURNS BOOLEAN AS (STARTS_WITH(LOWER(CAST(JSON_EXTRACT_SCALAR(summary, "$.mimeType") AS STRING)), 'image/'));
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics STRING,
+                            attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attr, "$.value") AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(JSON_EXTRACT(custom_metrics, '$.performance.lcp_elem_stats.attributes'))) AS attr
+  WHERE
+      JSON_EXTRACT_SCALAR(attr, "$.name") = attribute
+  LIMIT
+    1 ) );
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics STRING,
+                                attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.Images')) AS attributes
+  WHERE
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) IS NOT NULL
+    AND JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", 'url')) = GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
+                                                                                         'url')
+  LIMIT
+    1 ) );
+WITH
+  pagesWithLcpImages AS (
+    SELECT
+      date,
+      client,
+      page,
+      GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
+                                'url') AS url,
```

Comment:
I think this is a big part of the problem. The data for the LCP element is literally HTML attributes - so not `url` but `src` would need to be used here. I assume there are a few strange sites that have a `url` attribute on the image :)

Of course that's not a guarantee, but I believe just that change would already give us _a lot_ more data to work with.

Reply:
Dang 🤦  That must be it.

There are `$._performance.lcp_elem_stats.url` and `$_performance.lcp_elem_stats.attributes.src` — I must have mixed those up at some point

Reply:
Yep! Now I get tons of images included, also of all 6 MIME types. Good to see AVIF has more usage than BMP 😂

Reply:
> There are `$._performance.lcp_elem_stats.url` and `$_performance.lcp_elem_stats.attributes.src` — I must have mixed those up at some point

Does `$._performance.lcp_elem_stats.url` always have the actually loaded URL? In that case, maybe even better would be to compare that value with the `url` from the `Images` data.

Reply:
I would assume so, it is this value: https://developer.mozilla.org/en-US/docs/Web/API/LargestContentfulPaint/url

If my hunch is right (that I previously didn't try `performance.lcp_elem_stats.naturalWidth` but attempted to use performance.lcp_elem_stats.attributes.naturalWidth` instead, then maybe I don't even need `Images`. Testing now...

Otherwise, yes, that would be better to use for the comparison.